### PR TITLE
fix(util/git): Skip PR update on "force-with-lease" errors

### LINKED ISSUE #12124: PR gets auto-merged with changed version in title, but not changed contents
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Describe the bug

Our Dependency Dashboard [^dashboard] says:

> ##Ignored or Blocked
> 
> These are blocked by an existing closed PR and will not be recreated unless you click a checkbox below.
> 
> - [ ] chore(deps): update dependency yarn to v1.22.15

Sequence of events:

1. Renovate opens PR for yarn `v1.22.14` [^pr]
1. Renovate updates PR title to  `v1.22.15`
1. Renovate automerges with PR title `v1.22.15` but contents still at `v1.22.14`
1. Dashboard gets stuck because it thinks update `v1.22.15` is merged (it is, just with wrong file contents)

We still need to update to yarn `v1.22.15` in our `.devcontainer/Dockerfile` file:

https://github.com/renovatebot/renovate/blob/a09231f1af894a6fdb26064301030e905507d700/.devcontainer/Dockerfile#L5

[^dashboard]: https://github.com/renovatebot/renovate/issues/2958
[^pr]: https://github.com/renovatebot/renovate/pull/12055

### Relevant debug logs

I don't have access to the logs on the `renovatebot/renovate` repository, sorry!

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Wrap to `ExternalHostError` and re-throw, which will lead `processBranch()` to skip the branch updating for this run.

## Context:

- Closes: #12124

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/error.ts [hunk 1]
```diff
@@ -124,7 +124,7 @@ export function handleCommitError(
     logger.info(
       'Branch update was rejected because local copy is not up-to-date.'
     );
-    return null;
+    throw new ExternalHostError(err, 'git');
```
Comment (zharinov) [inline lib/util/git/error.ts:127]: `'[rejected] (stale info)'` can be added to `externalHostFailureStrings` instead, but log message would be too verbose IMO.

Anyway, please suggest better alternative if the solution is okay in general.

**Comment by zharinov** [comment]
Log messages from reproduction repo:

```
DEBUG: Pushing branch renovate/yarn-1.x (repository=renovate-testing/test-12124-pr-title-and-body-failure, branch=renovate/yarn-1.x)
 INFO: Branch update was rejected because local copy is not up-to-date. (repository=renovate-testing/test-12124-pr-title-and-body-failure, branch=renovate/yarn-1.x)
DEBUG: Closing Dependency Dashboard (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Retrieving issueList (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Retrieved 0 issues (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Removing any stale branches (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: config.repoIsOnboarded=true (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Branch lists (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "branchList": ["renovate/yarn-1.x"],
       "renovateBranches": ["renovate/yarn-1.x"]
DEBUG: remainingBranches= (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: No branches to clean up (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Repository timing splits (milliseconds) (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "splits": {"init": 2194, "extract": 1561, "lookup": 187, "update": 16636},
       "total": 20941
DEBUG: http statistics (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "urls": {
         "https://api.github.com/graphql (POST)": 4,
         "https://api.github.com/repos/renovate-testing/test-12124-pr-title-and-body-failure/contents/renovate.json (GET)": 1,
         "https://api.github.com/repos/renovate-testing/test-12124-pr-title-and-body-failure/pulls (GET)": 1
       },
       "hostStats": {"api.github.com": {"requestCount": 6, "requestAvgMs": 331, "queueAvgMs": 0}},
       "totalRequests": 6
 INFO: Repository finished (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "durationMs": 20941
DEBUG: Renovate exiting
```

File: lib/util/git/error.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/git/error.ts:127]: How does it look in the logs with your current suggestion?

Comment (zharinov) [reply lib/util/git/error.ts:127]: > How does it look in the logs with your current suggestion?

Please, take a look here: https://github.com/renovatebot/renovate/pull/14228#issuecomment-1039875472

**Comment by rarkins** [comment]
I was expecting to see an err logged somehow. Do we swallow it all now?

**Comment by zharinov** [comment]
```
DEBUG: Pushing branch renovate/yarn-1.x (repository=renovate-testing/test-12124-pr-title-and-body-failure, branch=renovate/yarn-1.x)
DEBUG: Converting git error to ExternalHostError (repository=renovate-testing/test-12124-pr-title-and-body-failure, branch=renovate/yarn-1.x)
       "err": {
         "task": {
           "commands": [
             "push",
             "origin",
             "renovate/yarn-1.x:renovate/yarn-1.x",
             "--force-with-lease",
             "-u",
             "--no-verify",
             "--verbose",
             "--porcelain"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "To https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git\n!\trefs/heads/renovate/yarn-1.x:refs/heads/renovate/yarn-1.x\t[rejected] (stale info)\nDone\nPushing to https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git\nerror: failed to push some refs to 'https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git'\n",
         "stack": "Error: To https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git\n!\trefs/heads/renovate/yarn-1.x:refs/heads/renovate/yarn-1.x\t[rejected] (stale info)\nDone\nPushing to https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git\nerror: failed to push some refs to 'https://github.com/renovate-testing/test-12124-pr-title-and-body-failure.git'\n\n    at Object.action (/Users/sergio/Work/renovate/node_modules/src/lib/plugins/error-detection.plugin.ts:38:28)\n    at PluginStore.exec (/Users/sergio/Work/renovate/node_modules/src/lib/plugins/plugin-store.ts:24:29)\n    at /Users/sergio/Work/renovate/node_modules/src/lib/runners/git-executor-chain.ts:114:40\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/Users/sergio/Work/renovate/node_modules/src/lib/runners/git-executor-chain.ts:111:14)\n    at GitExecutorChain.<anonymous> (/Users/sergio/Work/renovate/node_modules/src/lib/runners/git-executor-chain.ts:88:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/Users/sergio/Work/renovate/node_modules/simple-git/cjs/index.js:55:24)"
       }
DEBUG: Closing Dependency Dashboard (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Retrieving issueList (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Retrieved 0 issues (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Removing any stale branches (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: config.repoIsOnboarded=true (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Branch lists (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "branchList": ["renovate/yarn-1.x"],
       "renovateBranches": ["renovate/yarn-1.x"]
DEBUG: remainingBranches= (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: No branches to clean up (repository=renovate-testing/test-12124-pr-title-and-body-failure)
DEBUG: Repository timing splits (milliseconds) (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "splits": {"init": 2242, "extract": 1583, "lookup": 121, "update": 13675},
       "total": 17962
DEBUG: http statistics (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "urls": {
         "https://api.github.com/graphql (POST)": 4,
         "https://api.github.com/repos/renovate-testing/test-12124-pr-title-and-body-failure/contents/renovate.json (GET)": 1,
         "https://api.github.com/repos/renovate-testing/test-12124-pr-title-and-body-failure/pulls (GET)": 1
       },
       "hostStats": {"api.github.com": {"requestCount": 6, "requestAvgMs": 313, "queueAvgMs": 0}},
       "totalRequests": 6
 INFO: Repository finished (repository=renovate-testing/test-12124-pr-title-and-body-failure)
       "durationMs": 17962
DEBUG: Renovate exiting
```

**Comment by zharinov** [comment]
Ah, haven't noticed you're okay with swallowing. So yes, at first I decided we don't need to log it as we handle it now explicitly.
Which approach would you like to be released?

**Review by viceice — APPROVED** [review]

**Pull request merged**